compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/FoldConstantLowering.kt

```diff
@@ -0,0 +1,348 @@
+/*
+ * Copyright 2010-2018 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license
+ * that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.backend.jvm.lower
+
+import org.jetbrains.kotlin.backend.common.FileLoweringPass
+import org.jetbrains.kotlin.backend.common.phaser.makeIrFilePhase
+import org.jetbrains.kotlin.backend.jvm.JvmBackendContext
+import org.jetbrains.kotlin.builtins.KotlinBuiltIns
+import org.jetbrains.kotlin.builtins.PrimitiveType
+import org.jetbrains.kotlin.descriptors.CallableMemberDescriptor
+import org.jetbrains.kotlin.ir.declarations.IrFile
+import org.jetbrains.kotlin.ir.expressions.IrCall
+import org.jetbrains.kotlin.ir.expressions.IrConst
+import org.jetbrains.kotlin.ir.expressions.IrConstKind
+import org.jetbrains.kotlin.ir.expressions.IrExpression
+import org.jetbrains.kotlin.ir.expressions.impl.IrConstImpl
+import org.jetbrains.kotlin.ir.types.*
+import org.jetbrains.kotlin.ir.visitors.IrElementTransformerVoid
+import org.jetbrains.kotlin.ir.visitors.transformChildrenVoid
+import org.jetbrains.kotlin.name.FqName
+import org.jetbrains.kotlin.resolve.DescriptorUtils
+
+internal val foldConstantLoweringPhase = makeIrFilePhase(
+    ::FoldConstantLowering,
+    name = "FoldConstantLowering",
+    description = "Constant Folding"
+)
+
+/**
+ * A pass to fold constant expressions of most common types.
+ *
+ * For example, the expression "O" + 'K' + (1.toLong() + 2.0) will be folded to "OK3.0" at compile time.
+ */
+class FoldConstantLowering(private val context: JvmBackendContext) : IrElementTransformerVoid(), FileLoweringPass {
```

Comment:
This lowering pass adds quite some functionality to the JVM_IR back-end. I'd like to see some test for that.
Ideally, this should include two kinds of tests:
1. Box tests - to make sure everything works as expected;
2. Bytecode tests - to make sure that constant expressions are, indeed, folded.
Look into compiler/testData/codegen/box and compiler/testData/codegen/bytecodeText.

Pay attention to tests that can cause exceptions - e.g., division by zero.
```
fun test1() = 1 / 0 // ArithmeticException at run-time
fun test2() = 1.0 / 0.0 // Infinity
```

Also make sure that floating-point number comparisons work fine with respect to IEEE754 vs total order (various "special" numbers such as `-0.0` and `java.lang.Double.NaN`).

Reply:
It looks like that most of the box tests are already there in the code base? Although only either constant folding or generated bytecode will be tested, but not both; If an expression is folded, the bytecode generated won't be tested, and vice versa.

Since this is constant folding, the result should be a constant stored in the text. I'll start by adding bytecode tests and check whether there are box tests which are folded/optimized away.

Reply:
Please take a look at the new commit. They are added but not enabled yet; Because the current backed doesn't fold those comparisons, the new tests would fail on it. This [PR](https://github.com/JetBrains/kotlin/pull/2137) tries to fix that.